### ISSUE TITLE
Should "ltac: (idtac 123)." and "integer: (1)." be syntactically valid?

### ISSUE THREAD

**Issue by jfehrle** [coq/coq#12302]
If you run "Print Grammar tactic.", the last nonterminal it describes is `tactic_arg`, which includes 28 productions that are defined by OCaml code rather than listed in the `.mlg`s in the usual way.

Attaching them to `tactic_arg` permits the following syntactically valid but pointless sentences:
```
ltac: (idtac 123).  (* works, prints "123", but works without "ltac:(...)", too *)
integer: (1).       (* error: Expression does not evaluate to a tactic. *)
```

Should these productions have been attached to `tactic_arg_compat` instead?  And if so, should we change this?

Here are relevant productions from the `.mlg`s:
```
tactic_expr1: [
| tactic_arg
| reference LIST0 tactic_arg_compat
]

tactic_arg_compat: [
| tactic_arg
]

tactic_arg: [
| constr_eval
| "fresh" LIST0 fresh_id
| "type_term" uconstr
| "numgoals"
]
```

Also, I wonder why `tactic_arg_compat` ends with `compat`, which I expect is short for `compatibility`.  Is that relevant or unrelated?

**Comment by jfehrle** [comment]
Also, the production using `test_lpar_id_colon` is never syntactically valid because it's a lookahead-only symbol for for `( ident :`.

**Comment by jfehrle** [comment]
Searching all the .v files in the library and tests, I see only the following uses:
- constr - > 100 times
- open_constr - 1 time
- uconstr - 1 time

**Comment by JasonGross** [comment]
These constructs are not pointless, and are quite important.  Consider `let f x := x; x in f ltac:(idtac 123)`.  This does not work without `ltac:`.  The same is true of any tactic construct that needs a specific kind of argument.

As for constr, uconstr, and open_constr, they are all important (though uconstr can perhaps disappear in Ltac2).  The first typechecks it's argument and fills in holes, the second fills in holes with evars, and the third builds the term without filling in holes.

**Comment by jfehrle** [comment]
Read "sentences" instead of "constructs" at the top (which I updated).  These sentences (where the construct appears where a tactic name is expected) are pointless.  Your example has `f` as the tactic name, which is something different.  My question was:

> Should these productions have been attached to tactic_arg_compat instead?

**Comment by JasonGross** [comment]
Here are the ones that don't seem obviously about passing arguments to other tactics that I don't understand:
- glob_constr_with_bindings (parses constr_with_bindings) (what is this, and what is it for / used in?)
- casted_constr and glob (parses constr) - lconstr is a located constr, right?  Why does anything take non-located constrs, and what are these for?
- lglob (parses lconstr) - what is this? What uses it?

The other ones seem to either obviously be about binding arguments to pass to tactics, or else to be useful.

Though I'm not sure if all the argument kinds are useful.  For example, I seem to recall there being an "orient" one about rewrite directions, but can this actually be used with rewrite, or is it invalid because rewrite takes an optional orient, not an optional orient_or_var?

**Comment by jfehrle** [comment]
> As for constr, uconstr, and open_constr, they are all important

According to the `Print Grammar tactic` they're defined identically here, all referring to `constr.lconstr`.  `lglob` and  `lconstr` are also equivalent the other 3.

**Comment by jfehrle** [comment]
> I seem to recall there being an "orient" one about rewrite directions, but can this actually be used with rewrite

Yes, there is an `orient` and `rewrite` and others use `orient`.

**Comment by JasonGross** [comment]
> According to the Print Grammar tactic they're defined identically here, all referring to constr.lconstr. lglob and lconstr are also equivalent to the other 3.

The parsing is the same, the interpretation/semantics are different.

> Yes, there is an orient. rewrite and others use orient.

Yes, but does `let x := orient:(->) in rewrite x lem` work?  Does `let v := fresh in let x := ipattern:([v]) in intros x` work?  Does this work for other related things?

**Comment by ppedrot** [comment]
@jfehrle I am not sure about what you meant in your comment, are you complaining that these expressions can be used as a toplevel expression? If so, there are e.g. use cases of the form `let x := ltac:(blah) in bluh` to force Ltac to interpret some expressions in a particular way.

**Comment by jfehrle** [comment]
@ppedrot > are you complaining that these expressions can be used as a toplevel expression?

Yes, observing that they don't seem useful at the beginning of toplevel expressions.  Maybe only useful for the parameters of a tactic.

@JasonGross > does let x := orient:(->) in rewrite x lem work?

It appears it doesn't.  For example, I tried modifying app_eq_unit in `theories/Lists/List.v`:

```
    let foo := orient:(->) in rewrite foo E.  (* error shown below, "E" is highlighted in red *)
    foo; auto.   (* modified tactics *)

    rewrite -> E; auto.  (* original tactics *)
```

```
Syntax error: [tactic:ltac_use_default] expected after [tactic:tactic] (in [vernac:tactic_command]).
frame @ file "ide/protocol/xmlprotocol.ml", line 664, characters 29-44
frame @ file "ide/idetop.ml", line 442, characters 12-15
frame @ file "ide/idetop.ml", line 94, characters 8-64
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "lib/flags.ml", line 17, characters 14-17
raise @ file "clib/exninfo.ml", line 65, characters 2-15
frame @ file "vernac/vernacstate.ml", line 22, characters 10-35
raise @ file "clib/exninfo.ml", line 65, characters 2-15
```

The doc says "The evaluation yields either a term, an integer or a tactic." but `orient` is none of these.  That's true for most of the 28 possibilities.  

Your ipattern example gets an error, but this works in app_comm_cons (removing `[ ]`):

```
   let v := fresh "Z" in let x := ipattern:(v) in intros x.
```

`constr:` is used a bunch in the form `let x := constr:(...)`, that seems useful.  But looks like a lot of the 28 constructs don't work and shouldn't be mentioned in the doc and perhaps should go away.  Or made to work if there's sufficient benefit.  Does that make sense?

**Comment by ppedrot** [comment]
> The evaluation yields either a term, an integer or a tactic.

The documentation is wrong. The Ltac runtime contains many more toplevel values, e.g. as generated by generic argument.

> perhaps should go away

I don't think they should be removed. Even though Ltac is limited in its expressivity, there are still potential use cases for these quotations. For instance, if you want to make interact low-level ML code with Ltac.

**Comment by ppedrot** [comment]
> Maybe only useful for the parameters of a tactic.

Please don't make the language even less uniform than it is currently. Ltac has fancy beliefs about how you can manipulate data, e.g. discriminating between "lazy tactics" and "eager values". Disallowing let-bindings of quotations is likely to restrict what you can do with these values, since it's very hard to actually return values in Ltac. Usually you have to go through some form of CPS, see e.g how ring is implemented.

**Comment by jfehrle** [comment]
> The documentation is wrong.

Not surprising.  We'd like to get it right in the not-distant future.

> Please don't make the language even less uniform than it is currently.

Don't panic.  I'm trying to understand the use case.  Not much point in documenting things without understanding the use case.

> Usually you have to go through some form of CPS, see e.g how ring is implemented.

What is CPS?  `ring` is a Tactic Notation.  `constr_list` below is something else.

```
Tactic Notation (at level 0) "ring" :=
  let G := Get_goal in
  ring_lookup (PackRing Ring_gen) [] G.

Tactic Notation (at level 0) "ring" "[" constr_list(lH) "]" :=
  let G := Get_goal in
  ring_lookup (PackRing Ring_gen) [lH] G.
```

And what about `orient:( ... )` and `test_lpar_id_colon:( ... )`?  Where and how are those useful?

**Comment by ppedrot** [comment]
> What is CPS?

Continuation passing-style. If you look a few lines above this, you'll see higher-order functions used for the sake of emulating the return of values, *because Ltac*.
 
> And what about orient:( ... ) and test_lpar_id_colon:( ... )?

The first one is automatically introduced by the generic argument extension. It's probably not very useful in practice, since there are no way to use it other than pass it around until some ML function (and I don't think there is any that makes use of `orient` arguments, except `idtac`). For the second one, I have no idea, but I suspect it's Print Grammar not being clear enough about what argument this belongs to.

**Comment by Zimmi48** [comment]
Another way of asking Jim's question (taking into account PMP's concerns on making the language less uniform) is why does the syntax (in `tactic_expr1`) for a tactic applied to arguments, takes a `tactic_arg_compat` as argument instead of just a `tactic_expr0`? And why is there an indirection from `tactic_arg_compat` to `tactic_arg`?

@JasonGross AFAIK the difference between `constr` and `lconstr` is not the one that you said ("located constr"). Even @herbelin did not know why `lconstr` was called `lconstr`. The difference is that if you expect a `constr` and want to pass a term applied to arguments, then you'll need parentheses, which you won't need in the case of an `lconstr`. Cf. the explanation in the doc: https://coq.github.io/doc/master/refman/language/core/basic.html#grammar-token-term (`lconstr` is `term` and `constr` is `one_term`).

**Comment by ppedrot** [comment]
> why does the syntax (in tactic_expr1) for a tactic applied to arguments, takes a tactic_arg_compat as argument instead of just a tactic_expr0

Aha. This is the kind of question that makes me cry. The answer is "historical reasons", i.e. at some point it was decided that when a user wrote a Ltac snippet `tac foo`, then for *convenience* `foo` should be interpreted first as a Coq term, and in case of *parsing failure*, backtrack to trying to parse it as Ltac. If you change that you virtually break any single Ltac development.

Note that the restriction on the applicand is also absurd, in so far as it requires a *tactic reference*, not a *tactic expression*. But I am afraid that Ltac is overall nonsensical, not only semantic-wise, but also syntax-wise.

**Comment by Zimmi48** [comment]
Thanks for the explanation!

**Comment by jfehrle** [comment]
> AFAIK the difference between constr and lconstr is not the one that you said ("located constr").

That's correct in the .mlgs, which is relevant for #11761.  But confusingly, it appears that the `constr:(...)` and `lconstr:( ... )` both map to lconstr in the grammar.

**Comment by JasonGross** [comment]
Yes, this is because there is already parentheses around `constr:(...)`, so if it mapped to `constr` rather than `lconstr`, then you'd need to write `constr:((id foo))`.

**Comment by JasonGross** [comment]
Hence I don't think `lconstr:(...)` is useful, as it's a duplicate of `constr:(...)`.

**Comment by jfehrle** [comment]
I think it makes sense to document only `ltac:( ... )` and `constr:( ... )` for now.  We _might_ include a note saying there are others, some of which appear not to be useful.  And direct users to look at `Print Grammar tactic` to see the others.

> [orient and test_lpar_ident_colon are] probably not very useful in practice, since there are no way to use it other than pass it around until some ML function

This sounds like such a corner case--it would be in a plugin?--that we could safely deprecate these (and some others) and later remove them.  Removing useless, confusing stuff is a good thing even if ltac1 will never be truly clean.  You can't even get the test_lpar_ident_colon to parse!  But not today.

FWIW, `Tactic Notation` supports orient (strange):

```
Tactic Notation "foo" orient(x) := idtac x.
foo ->.   (* true *)
```

> If you change that you virtually break any single Ltac development.

Antiquoting in Ltac2 avoids this problem?  Conceptually, is it possible to create a script that could figure out where this hack was used and generate a modified script that would not rely on it?  Or just convert it to Ltac2?  I expect people will use Ltac1 for a very long time if the conversion has to be done by hand.

**Comment by JasonGross** [comment]
> I think it makes sense to document only `ltac:( ... )` and `constr:( ... )` for now. We _might_ include a note saying there are others, some of which appear not to be useful. And direct users to look at `Print Grammar tactic` to see the others.

No, `open_constr:(...)` and `uconstr:(...)` are quite important for writing powerful and performant automation in Ltac1.  In particular, `uconstr:(...)` is essential for building large terms while delaying their typechecking, as well as for passing around patterns with holes in them without allocating evars.  By contrast, `open_constr:(...)` is quite essential for creating evars without needing to jump through hoops to create them via `evar`, and is also quite essential for managing typeclass resolution and is quite useful for interacting with `simple refine` et al to control which evars turn into goals and which ones remain on the shelf.  It seems fine to document just these four, though I think `ident:(...)`, `reference:(...)`, `integer:(...)`, and `string:(...)`  are occasionally useful.  I wouldn't even now how to use the others without digging, though, and it seems fine to not document them.

**Comment by jfehrle** [comment]
> No, `open_constr:(...)` and `uconstr:(...)` are quite important for writing powerful and performant automation in Ltac1.

According to `Print Grammar tactic`, their syntax is identical to `lconstr:(...)` or `constr:(...)`, so they're actually not useful, either.

`ident`, `reference`, `integer` and `string`, yeah, I guess.  `reference` is just an `ident` though it looks like it preserves the `loc` information.  Not clear how that matters outside of ML code.

**Comment by JasonGross** [comment]
> Antiquoting in Ltac2 avoids this problem?

Yes

> Conceptually, is it possible to create a script that could figure out where this hack was used and generate a modified script that would not rely on it?

Generating a modified script is quite easy and at the same time quite ugly.  Every single application of a tactic to a constr would need to wrap the constr in `constr:(...)`, or for lighter impact, every single application of a tactic to another tactic would need to wrap the tactic in `ltac:(...)`.  Figuring out where it's used is significantly harder, but it shouldn't be too hard to write a flag that makes `Print Ltac` emit extra wrappers around `constr` and `ltac` things in these positions.

> Or just convert it to Ltac2? I expect people will use Ltac1 for a very long time if the conversion has to be done by hand.

That'd be pretty cool!  I think Ltac2 needs to be more feature-complete before we can hope for this, though, e.g., it needs a printf / better idtac.

**Comment by JasonGross** [comment]
> > No, `open_constr:(...)` and `uconstr:(...)` are quite important for writing powerful and performant automation in Ltac1.
>
> According to `Print Grammar tactic`, their syntax is identical to `lconstr:(...)` or `constr:(...)`, so they're actually not useful, either.

The *parsing syntax* is identical, but the *semantics* are quite different.
```coq
Goal True.
  let x := uconstr:(_) in idtac x. (* _ *)
  let x := open_constr:(_) in idtac x. (* ?y *)
  let x := constr:(_) in idtac x. (* Error: Cannot infer an internal placeholder of type "Type". *)
```

**Comment by JasonGross** [comment]
> `reference` is just an `ident` though it looks like it preserves the `loc` information. Not clear how that matters outside of ML code.

Once again, the syntax may be the same, but the semantics are quite different:
```coq
Goal True.
  let x := ident:(i) in idtac x. (* i *)
  let x := reference:(i) in idtac x. (* Error: The reference i was not found in the current environment. *)
```
and
```coq
Goal True.
  let x := ident:(I) in let x := constr:(fun x : Set => x) in idtac x. (* (fun I : Set => I) *)
  let x := reference:(I) in let x := constr:(fun x : Set => x) in idtac x. (* Error: Ltac variable x is not bound to an identifier.
It cannot be used in a binder. *)
```

**Comment by JasonGross** [comment]
> According to `Print Grammar tactic`, their syntax is identical to `lconstr:(...)` or `constr:(...)`, so they're actually not useful, either.

To me, this seems kind-of like saying that it's not useful to have both `pose` and `pose proof`, or to have both `match goal` and `match reverse goal`, because the syntax of these is the same.

**Comment by jfehrle** [comment]
Not at all.  `Print Grammar tactic` clearly suggests the _semantics_ are the same.  If they behave differently, then this rules are incorrect.  Here are the things where the lhs doesn't match the parenthesized nonterminal:

```
  | IDENT "glob_constr_with_bindings";  ":";  "("; tactic:constr_with_bindings;  ")"
  | IDENT "casted_constr";  ":";  "("; constr:constr;  ")"
  | IDENT "lglob";  ":";  "("; constr:lconstr;  ")"
  | IDENT "glob";  ":";  "("; constr:constr;  ")"
  | IDENT "ltac";  ":";  "("; tactic:tactic_expr LEVEL "5";  ")"
  | IDENT "open_constr";  ":";  "("; constr:lconstr;  ")"
  | IDENT "ipattern";  ":";  "("; tactic:simple_intropattern;  ")"
  | IDENT "constr";  ":";  "("; constr:lconstr;  ")"
  | IDENT "uconstr";  ":";  "("; constr:lconstr;  ")"
```
Versus, say `  | IDENT "occurrences";  ":";  "("; tactic:occurrences;  ")"` which is consistent.

@ppedrot if this is reasonably easy to fix and you give me instructions on what to change, I will put in a PR.  Of course, it would be even better to define these as regular productions in the .mlg where's they're no buried in the code.

**Comment by JasonGross** [comment]
> `Print Grammar tactic` clearly suggests the _semantics_ are the same. If they behave differently, then these rules are incorrect.

I don't understand, and am getting out of my depth in parsing.  I know that each grammar entry has three parts in the type signature, though I don't recall what they are.  Anyway, what about the rules suggests they have the same semantics to you?  I read them as just about parsing and not about interpretation.

**Comment by jfehrle** [comment]
_Incorrectly_ suggests, at least to me.  You showed they behave differently.  Let's leave it at that.

**Comment by jfehrle** [comment]
Do you have an example where `reference` is accepted and different from using `ident`?  Both your examples above failed for `reference`.  (Not necessarily planning to include such as example in the doc.)

How about cases where `integer` and `string` are necessary, not optional?

**Comment by ppedrot** [comment]
> so they're actually not useful

I think that if you only consider things from the angle of syntax, you're having a *really* partial understanding of the situation, notwithstanding the fact that Print Grammar is probably buggy in its printing. As @JasonGross said, these constructions have different semantics.

> Print Grammar tactic clearly suggests the semantics are the same.

What did you infer this from?

> Do you have an example where reference is accepted and different from using ident?

```
let x := reference:(Peano.plus) in unfold x.
```

> How about cases where integer and string are necessary, not optional?

```
let x := string:("foo") in idtac x.
```

```
let f := fun x => do x idtac "foo" in
f integer:(0).
```

**Comment by Zimmi48** [comment]
@jfehrle This seems completely absurd to think that parsing should determine semantics. If `Print Grammar` only prints the parsing rules, where is the problem? This doesn't say anything about the semantics.

**Comment by Matafou** [comment]
After being helped by @JasonGross I found "lconstr" pretty useful. It probably means "List of CONSTR, interpreted as an application". It should be documented IMHO.

**Comment by jfehrle** [comment]
@Matafou I'll put it in.  `lconstr` is syntactically equivalent to `operconstr` level 200 in the mlg, which allows application, whereas `constr` is equivalent to level 8, in which application needs to be parenthesized.

**Comment by JasonGross** [comment]
I think `lconstr` is quite useful for `Tactic Notation` (assuming it's supported) and `TACTIC EXTEND`, but I think `lconstr:(...)` is useless, being both syntactically and semantically identical to `constr:(...)`.

**Comment by JasonGross** [comment]
@Matafou where did you get the interpretation "list of constr" from?  I thought @herbelin and others did not recall what it stood for (though "list of constr") makes more sense than anything else I've heard.

**Comment by jfehrle** [comment]
> I think that if you only consider things from the angle of syntax, you're having a *really* partial understanding of the situation, notwithstanding the fact that Print Grammar is probably buggy in its printing.

Certainly not limiting myself to syntax--but syntactic clues are relatively easy for me to find, while the semantic differences could be extremely hard to deduce from the code or examples.
Figuring out how this stuff works involves a lot of guesswork and experimentation.

> > Print Grammar tactic clearly suggests the semantics are the same.

> What did you infer this from?

I said "suggests", not "inferred".  From the mismatch between the IDENT and the nonterminal within the parentheses.  Looking at the grammar, one might expect these constructs would be thin wrappers around the nonterminals `constr` and `uconstr` rather than `lconstr`--not certain but a reasonable guess:

```
  | IDENT "constr";  ":";  "("; constr:lconstr;  ")"
  | IDENT "uconstr";  ":";  "("; constr:lconstr;  ")"
```

Thanks for the examples.  I notice that `string(...)` doesn't work for this example: `let x := string:("abc") in let y := fresh x in idtac y.`, so `string` may not be useful other than in `idtac`.  (`idtac` will print an `orient(...)` value, but AFAIK `orient` isn't really usable for, say, `rewrite`--the grammar doesn't allow for an identifier in place of `->` or `<-`.)

**Comment by jfehrle** [comment]
> This seems completely absurd to think that parsing should determine semantics.

But I didn't say that.  Grammars may document semantics a bit.  For example, `g_tactic.mlg` has these productions:

```
  open_constr:
    [ [ c = constr -> { c } ] ]
  ;
  uconstr:
    [ [ c = constr -> { c } ] ]
  ;
```

Since the parser action just copies the `constr` value, they add no syntax.  Presumably they were created to help document which semantics are applied to the `constr` in various productions.  For example, you would expect `| "replace" uconstr clause` will be applying semantics different from where `open_constr` appears.

> If Print Grammar only prints the parsing rules, where is the problem? This doesn't say anything about the semantics.

See my previous comment mentioning "thin wrappers".

**Comment by ppedrot** [comment]
I did a bit of archeology, and the `lconstr` name was introduced in 46e708f92deef78043f4f221293df131e29aeff1 from the renaming of `lcommand`. Tracking the latter, it seems to have meant indeed "list of commands" as far as the Coq 4.X branches, and slowly evolved over time when seemingly commands were separated from terms.

**Comment by jfehrle** [comment]
> I think `lconstr` is quite useful for `Tactic Notation` (assuming it's supported) and `TACTIC EXTEND`, but I think `lconstr:(...)` is useless, being both syntactically and semantically identical to `constr:(...)`.

Sigh.  They are syntactically different things in the grammar.  It's perverse that `constr:(...)` actually accepts the syntax for `lconstr`.  A bug or maybe intentional to make it "easier" to use an undocumented feature.

**Comment by ppedrot** [comment]
> It's perverse

I don't see the perversity, but I do see the uniformity. The rule is that any generic argument defines both a tactic notation scope *and* and an Ltac quotation entry which parses this scope. What I find perverse is the fact that one has to define a new generic argument to merely change the level at which the corresponding entry is parsed, and that `lconstr` shouldn't exist at all.

**Comment by jfehrle** [comment]
> I don't see the perversity, but I do see the uniformity.

The name in constr(...) doesn't match the underlying construct.  I think this is perverse ("contrary to the accepted or expected standard or practice").

> The rule is that any generic argument defines both a tactic notation scope and and an Ltac quotation entry which parses this scope.

Meaning that they should support the same set of items?
`int_or_var` is not accepted as a quotation entry:

```
Tactic Notation "foo" int_or_var(a) := idtac a.
foo 3.   (* 3 *)
let x := int_or_var:(1) in idtac x.  (* syntax error (at the colon) *)
```

Thanks for all your comments.  I think I'm done with this for now.
